File: baseball/src/tool/test/proto/presence.cs
using System;
using System.Collections.Generic;
//using System.Linq;
using System.Text;
using ProtoBuf;
//using Google.ProtocolBuffers;

namespace nPROTO
{

    public enum PresenceCommand {
	    NONE = 0,
	    LOGIN = 1,
	    LOGOUT = 2,
	    GAMEENTER = 3,
	    GAMELEAVE = 4,
	    DETACH = 5,
	    PLUG = 6,
	    CATEGORY = 7,
	    ROOMLIST = 8
    }

    [ProtoContract]
	public class PReqProtocol
    {
        [ProtoMember(1, IsRequired = true)]
		public PresenceCommand command {get;set;}
        [ProtoMember(2, IsRequired = true)]
		public int seqID {get;set;}
        [ProtoMember(3, IsRequired = true)]
		public int gameID {get;set;}
        [ProtoMember(4, IsRequired = true)]
		public int svcID {get;set;}
        [ProtoMember(5, IsRequired = true)]
		public int category {get;set;}

	    [ProtoMember(6, IsRequired = false)]
	    public int arg1 {get;set;}
	    [ProtoMember(7, IsRequired = false)]
	    public int arg2 {get;set;}
    }

    [ProtoContract]
	public class PAnsProtocol
    {
        [ProtoMember(1, IsRequired = true)]
		public PresenceCommand command{get;set;}
        [ProtoMember(2, IsRequired = true)]
		public int seqID {get;set;}
        [ProtoMember(3, IsRequired = true)]
		public int gameID {get;set;}
        [ProtoMember(4, IsRequired = true)]
		public int svcID {get;set;}

        [ProtoMember(5, IsRequired = true)]
		public int category {get;set;}

	    // true - success/false - fail
        [ProtoMember(6, IsRequired = true)]
		public bool result {get;set;}

        [ProtoMember(7, IsRequired = true)]
		public string status {get;set;}
        [ProtoMember(8, IsRequired = true)]
		public bool detached {get;set;}

	    [ProtoMember(9, IsRequired = false)]
    	public string ip {get;set;}
	    [ProtoMember(10, IsRequired = false)]
	    public int port {get;set;}
	    [ProtoMember(11, IsRequired = false)]
	    public int roomID {get;set;}
	    [ProtoMember(12, IsRequired = false)]
	    public int arg1 {get;set;}
	    [ProtoMember(13, IsRequired = false)]
	    public int arg2 {get;set;}
	    [ProtoMember(14, IsRequired = false)]
	    public string info {get;set;}
    }



}

File: baseball/src/server/trunk/gateway/tests/presence.cpp
#include "presence.hpp"

#include <netdb.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>

#define ENABLE_MSG_TRACE
#include "protocol/dump.hpp"

#include "protocol/presence.pb.h"

#include <google/protobuf/io/zero_copy_stream_impl.h>

using namespace std;
using google::protobuf::io::ArrayInputStream;


/* -------------------------------------------------------------------------- */

Presence::Presence():
  Server()
{
}

/* -------------------------------------------------------------------------- */

Presence::~Presence()
{
}

/* -------------------------------------------------------------------------- */

Server::Session* 
Presence::create_session()
{
  return new Presence::GSession();
}

/* -------------------------------------------------------------------------- */

bool
Presence::GSession::dispatch(unsigned char* packet, std::size_t length)
{
  cerr << "PRESENCE\n";

  ArrayInputStream input(packet, length);

  PReqProtocol request;

  if (!request.ParseFromZeroCopyStream(&input))
  {
    return false;
  }
  
  MSG::DumpPReqProtocol(request);

  PAnsProtocol answer;

  switch(request.command())
  {
  case LOGIN:
    {
      answer.set_command(LOGIN);
      answer.set_seqid(request.seqid());
      answer.set_gameid(request.gameid());
      answer.set_svcid(request.svcid());
      answer.set_category(request.category());
      answer.set_result(true);
      answer.set_status("L");
      answer.set_detached(false);
      answer.set_ip("");
      answer.set_port(0);
      answer.set_roomid(0);
    }
    break;
  case LOGOUT:
    {
      answer.set_command(LOGOUT);
      answer.set_seqid(request.seqid());
      answer.set_gameid(request.gameid());
      answer.set_svcid(request.svcid());
      answer.set_category(request.category());
      answer.set_result(true);
      answer.set_status("U");
      answer.set_detached(false);
      answer.set_ip("");
      answer.set_port(0);
      answer.set_roomid(0);
    }
    break;
  case GAMEENTER:
    {
      answer.set_command(GAMEENTER);
      answer.set_seqid(request.seqid());
      answer.set_gameid(request.gameid());
      answer.set_svcid(request.svcid());
      answer.set_category(request.category());
      answer.set_result(true);
      answer.set_status("G");
      answer.set_detached(false);
      answer.set_ip("localhost");
      answer.set_port(6020);
      answer.set_roomid(10);
    }
    break;
  case GAMELEAVE:
    {
      answer.set_command(GAMELEAVE);
      answer.set_seqid(request.seqid());
      answer.set_gameid(request.gameid());
      answer.set_svcid(request.svcid());
      answer.set_category(request.category());
      answer.set_result(true);
      answer.set_status("L");
      answer.set_detached(false);
      answer.set_ip("");
      answer.set_port(0);
      answer.set_roomid(0);
    }
    break;
  case DETACH:
    return false;
  case PLUG:
    return false;
  case CATEGORY:
    {
      answer.set_command(CATEGORY);
      answer.set_seqid(request.seqid());
      answer.set_gameid(request.gameid());
      answer.set_svcid(request.svcid());
      answer.set_category(request.category());
      answer.set_result(true);
      answer.set_status("L");
      answer.set_detached(false);
      answer.set_ip("");
      answer.set_port(0);
      answer.set_roomid(0);
      answer.set_info("10,20,30,40,50");
    }
    break;

  default:
    return false;
  }

  MSG::DumpPAnsProtocol(answer);
  send(answer);

  return true;  
}

/* -------------------------------------------------------------------------- */


